prov/gni: Use htd_buf  as intermediate buffer for alignment

a-abraham opened the pull request:
Move htd_buf from fab_req.msg to fab_req

Fixes ofi-cray/libfabric-cray#1087

Now that htd_bufs are available on the top level fab_req
utilize that as the intermediate buffer for chained gets etc.
and remove the int_buf.

ofi-cray/libfabric-cray#1077

Signed-off-by: Amith Abraham <aabraham@cray.com>

@ztiffany  @sungeunchoi

a-abraham commented:
I have a question in regard to these changes -  are what are the implications of having the following? GNIX_HTD_POOL_SIZE 128
Any reason we might want to increase the size? 
@ztiffany

ztiffany commented:
We need to change these htd bufs to be a growable pool eventually.  128 probably won't be limiting for now.

ztiffany commented on prov/gni/src/gnix_msg.c:
```diff
@@ -962,6 +963,16 @@ static int __gnix_rndzv_req(void *arg)
+			req->htd_buf_e = _gnix_ep_get_htd_buf(ep);
+			if (req->htd_buf_e == NULL)
+				GNIX_WARN(FI_LOG_EP_DATA,
+					  "RAN OUT OF HTD_BUFS");
```
Need to return error here.

a-abraham replied:
What would the return code need to be?

ztiffany replied:
Use -FI_ENOSPC to match the tx_alloc failure right above.

a-abraham replied:
Fixed.

sungeunchoi commented:
👍  other than changing `htd` to something that makes more sense in the current context.

ztiffany commented on prov/gni/src/gnix_rma.c:
```diff
@@ -511,6 +511,16 @@ static void __gnix_rma_fill_pd_chained_get(struct gnix_fab_req *req,
+	if (req->htd_buf_e == NULL) {
+		req->htd_buf_e = _gnix_ep_get_htd_buf(ep);
+		if (req->htd_buf_e == NULL)
+			GNIX_WARN(FI_LOG_EP_DATA, "RAN OUT OF HTD_BUFS");
```
Return error.

a-abraham replied:
Since this a static void function,  should I just use an assert statement?

a-abraham replied:
oh nvm I just saw Sung's comment

ztiffany commented on prov/gni/src/gnix_rma.c:
```diff
@@ -573,10 +580,20 @@ static void __gnix_rma_fill_pd_indirect_get(struct gnix_fab_req *req,
+	if (req->htd_buf_e == NULL) {
+		req->htd_buf_e = _gnix_ep_get_htd_buf(ep);
+		if (req->htd_buf_e == NULL)
+			GNIX_WARN(FI_LOG_EP_DATA, "RAN OUT OF HTD_BUFS");
```
Return error.

ztiffany commented on prov/gni/src/gnix_rma.c:
```diff
@@ -642,6 +659,15 @@ int _gnix_rma_post_rdma_chain_req(void *data)
+	if (req->htd_buf_e == NULL) {
+		req->htd_buf_e = _gnix_ep_get_htd_buf(ep);
+		if (req->htd_buf_e == NULL)
+			GNIX_WARN(FI_LOG_EP_DATA, "RAN OUT OF HTD_BUFS");
```
Return error.

a-abraham replied:
Fixed

a-abraham commented:
In gnix_msg.c   line 1251 onwards are all super indented and way over the Line limit of 80 characters. 

For some of the lines that I will edit in that section (just to rename HTD to INT_TX), it's not super easy to get the lines under the 80 character limit, unless I reformat most of the surrounding code...  Is it okay if I leave the line lengths as is and write up a separate issue to address those?
@sungeunchoi

a-abraham commented on prov/gni/src/gnix_rma.c:
```diff
@@ -511,6 +511,16 @@ static void __gnix_rma_fill_pd_chained_get(struct gnix_fab_req *req,
+	if (req->htd_buf_e == NULL) {
+		req->htd_buf_e = _gnix_ep_get_htd_buf(ep);
+		if (req->htd_buf_e == NULL)
+			GNIX_WARN(FI_LOG_EP_DATA, "RAN OUT OF HTD_BUFS");
```
Fixed

sungeunchoi commented:
Yeah don't worry about line spacing too much if it doesn't make sense to stay with 80 chars.

sungeunchoi commented:
Check the LANL CI build log.  You might need to return error type on some functions.

a-abraham commented:
I've returned Errors when we run out of int_tx buffers where applicable. In other areas I've added GNIX_FATAL messages along with TODO statements explaining that the pools will be growable in the future.

I've also added the third commit which renames htd_bufs  to int_tx bufs.  Let me know if there's anything else you notice that needs attention - @ztiffany @sungeunchoi

ztiffany commented:
:+1:
